File: CV_tools.py
#!/usr/bin/env python

import json
import sys

sections_names = ["section*", "subsection*", "subsubsection*"]

def escape(f):
    return f.replace("\n", "\\\\").replace("^", "\\^{}").replace("#", "\\#").replace("%", "\\%")


def handle_children(js, level):
    for s in js["sections"]:
        if "type" not in s:
            s["type"] = "None"

        if s["type"] == "employment":
            print("\\end{multicols}")

        if "name" in s:
            print("\\" + escape(sections_names[level]) + "{" + s["name"] + "}")

        if len(s["type"]) > 6 and s["type"][:6] == "result":
            print("\\textit{"+s["when"]+", "+s["where"]+"}")


        if s["type"] == "column_break":
            print("\\vfill\\null\n\\columnbreak")
        elif s["type"] == "h_rule":
            print("\\end{multicols}")
            print("\\rule{\\textwidth}{1pt}")
            print("\\begin{multicols}{2}")
        elif s["type"]=="text":
            print(s["data"])
        elif s["type"] == "employment":
            print("\\begin{multicols*}{2}")
            for d in s["data"]:
                print("\\textbf{"+d["when"]+"}")
                print(d["name"]+", "+d["where"]+". "+d["more"]+"\\par")
            print("\\end{multicols*}")
            print("\\begin{multicols}{2}")
        elif s["type"] == "results_vtable":
            print("\\vspace{1em}\\par\\begin{center}\\begin{tabular}{ "+"c "*s["cols"]+"}")
            q = 0
            for f in s["data"]:
                q+=1
                t ="\\makecell{"+escape(f)+"\\vspace{0.5em}}"
                if q == s["cols"]:
                    q=0
                    print(t+"\\\\")
                else:
                    print(t+"&")
            if q!=0:
                print("& "*(s["cols"]-q-1)+" \\\\")
            print("\\end{tabular}\\end{center}")
        elif s["type"] == "results_htable":
            print("\\vspace{1em}\\par\\begin{center}\\begin{tabular}{ "+"c "*s["cols"]+"}")
            q = 0
            r_1 = ""
            r_2 = ""
            for _s,_g in s["data"].items():
                q+=1
                a ="\\makecell{\\textbf{"+escape(_s)+"}}"
                b ="\\makecell{"+escape(_g)+"\\vspace{0.5em}}"
                if q == s["cols"]:
                    q=0
                    r_1+=a+"\\\\"
                    r_2+=b+"\\\\"
                    print(r_1)
                    print(r_2)
                    r_1=""
                    r_2=""
                else:
                    r_1+=a+"& "
                    r_2+=b+"& "
            if q!=0:
                r_1+="& "*(s["cols"]-q-1)+" \\\\"
                r_2+="& "*(s["cols"]-q-1)+" \\\\"
                print(r_1)
                print(r_2)
                r_1=""
                r_2=""
            print("\\end{tabular}\\end{center}")
        elif s["type"] == "list":
            print("\\begin{itemize}")
            for v in s["data"]:
                print("\\item")
                if "head" in v:
                    print("\\textbf{"+escape(v["head"])+":} ")
                if "data" in v:
                    print(escape(v["data"]))
                if "url" in v:
                    print("\\url{"+escape(v["url"])+"}")
            print("\\end{itemize}")
        else:
            #print(s["type"])
            pass
        if "sections" in s:
            handle_children(s, level + 1)


js = None
exe, path = sys.argv
with open(path, "r") as file:
    js = json.loads(file.read())

print(
    """\\documentclass{article}
\\usepackage[utf8]{inputenc}
\\usepackage{auto-pst-pdf}
\\usepackage{fontenc,unicode-math}
\setmainfont[Ligatures=TeX]{TeX Gyre Adventor}

\\usepackage{multicol}
\\usepackage{makecell}
\\usepackage{geometry}
\\usepackage{url}
\\geometry{a4paper, portrait, margin=1.5cm}

\\begin{document}
"""
)
print("\\begin{center}")
print("\\textbf{\\LARGE "+js["name"] + "}\\\\")
print(js["email"])
print("\\end{center}")
print(js["desc"])
print("\\begin{multicols}{2}")
handle_children(js, 0)
print("\\end{multicols}")
print("\\end{document}")

File: CV_tools.php
<?php

function handle_children($json, $level)
{
    foreach ($json["sections"] as $value)
    {
        if (isset($value["name"]))
        {
            echo "<h$level>{$value["name"]}</h$level>\n";
        }
        if(!isset($value["type"]))
        {
          $value["type"]="";
        }

        if ($value["type"] == "column_break" or $value["type"] == "h_rule")
        {
            // do nothing, this is for LaTeX typesetting
        }
        elseif ($value["type"]=="text") {
          echo $value["data"];
        }
        elseif ($value["type"]=="results_vtable") {
          echo "<span class='when'>{$value["when"]}, {$value["where"]}</span>";
          echo "<table>\n\t<tr>";
          for($i=0;$i< count($value["data"]);$i++)
          {

            $j =$i % $value["cols"];
            if($j== 0){
              echo "\t</tr>\n\t<tr>\n";
            }
            $t = nl2br($value["data"][$i]);

            echo "\t\t<td>{$t}</td>\n";
          }
          echo "\t</table>\n</table>";
        }
        elseif ($value["type"]=="results_htable") {
          echo "<span class='when'>{$value["when"]}, {$value["where"]}</span>";
          echo "<table>\n";
          $k = array_keys($value["data"]);
          for($i=0;$i< count($k);$i+=$value["cols"])
          {
            echo "\t<tr>\n";
            for ($j=0; $j < $value["cols"]; $j++) {
              echo "\t\t<th>{$k[$i+$j]}</th>\n";
            }
            echo "\t</tr>\n\t<tr>\n";
            for ($j=0; $j < $value["cols"]; $j++) {
              echo "\t\t<td>{$value["data"][$k[$i+$j]]}</td>\n";
            }
            echo "\t</tr>\n";
          }
          echo "</table>";
        }
        elseif ($value["type"] == "employment")
        {
          echo "<ul>\n";
          foreach ( $value["data"] as $var2) {
            echo "\t<li>";
            if(isset($var2["name"]))
            {
              echo "<b>{$var2["name"]}</b>";
            }
            if(isset($var2["where"]))
            {
              echo ", {$var2["where"]}";
            }
            if(isset($var2["when"]))
            {
              echo " ({$var2["when"]})";
            }
            if(isset($var2["more"]))
            {
              echo "({$var2["more"]})";
            }
            echo "</li>\n";
          }
          echo "</ul>\n";
        }
        elseif ($value["type"] == "list")
        {
          echo "<ul>\n";
          foreach ( $value["data"] as $var2) {
            echo "\t<li>";
            if(isset($var2["head"]))
            {
              echo "<b>{$var2["head"]}:</b> ";
            }
            if(isset($var2["data"]))
            {
              echo $var2["data"];
            }
            if(isset($var2["url"]))
            {
              echo " <a href='{$var2["url"]}'>{$var2["url"]}</a>";
            }
            echo "</li>\n";
          }
          echo "</ul>\n";
        }
        else
        {
            echo $value["type"];
            if (isset($value["data"]))
            {
                var_dump($value["data"]);
            }
        }
        if (isset($value["sections"]))
        {
            handle_children($value, $level + 1);
        }

    }
}

function handle_CV($p)
{
    $string = file_get_contents($p);
    $json_a = json_decode($string, true);
    echo "<header>\n";
    echo "\t<span class='name'>{$json_a["name"]}</span>\n";
    echo "\t<span class='email'>{$json_a["email"]}</span>\n<br><br>\n";
    echo "\t<span class='desc'>{$json_a["desc"]}</span>\n";
    echo "</header>\n";

    handle_children($json_a, 1);
}

?>
